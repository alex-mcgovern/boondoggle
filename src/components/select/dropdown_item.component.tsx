import clsx from "clsx";
import React, { forwardRef } from "react";

import { Box } from "../box";
import * as styles from "./select.styles.css";

import type { SharedUiScale } from "../..";
import type { DropdownItemShape } from "./select.types";
import type { Ref } from "react";

export interface DropdownItemProps {
  isHighlighted: boolean;
  isMultipleSelectionEnabled?: boolean;
  size?: SharedUiScale;
  isMulti?: boolean;
  item: DropdownItemShape;
  isDropdownItemSelected?: boolean;
}

export const DropdownItem = forwardRef(
  (
    {
      item,
      size,
      isHighlighted,
      isMulti,
      isDropdownItemSelected,
      ...rest
    }: DropdownItemProps,
    ref
  ) => {
    return (
      <Box
        as="button"
        id={item.label}
        className={clsx(
          styles.getDropdownItemStyles({
            size,
          }),
          {
            [styles.isHighlighted]: isHighlighted,
            [styles.isSelected]: isDropdownItemSelected,
          }
        )}
        {...rest}
        ref={ref as Ref<HTMLButtonElement>}
      >
        <Box flexShrink="0">{item.label}</Box>
        {isMulti && (
          <Box
            as="input"
            type="checkbox"
            readOnly
            tabIndex={-1}
            checked={isDropdownItemSelected}
            marginLeft="auto"
          />
        )}
      </Box>
    );
  }
);
