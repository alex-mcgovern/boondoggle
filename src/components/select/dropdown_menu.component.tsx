import { extractAtomsFromProps } from "@dessert-box/core";
import clsx from "clsx";
import React, { forwardRef } from "react";

import { getSprinkles } from "../../styles/utils/get_sprinkles.css";
import { Box } from "../box_component";
import { DropdownItem } from "./dropdown_item.component";
import * as styles from "./select.styles.css";

import type { SharedUiScale } from "../../styles/common/globalVariantsUiScale.css";
import type { SprinklesArgs } from "../../styles/utils/get_sprinkles.css";
import type { DropdownItemShape } from "./select.types";
import type {
  UseComboboxPropGetters,
  UseMultipleSelectionActions,
  UseMultipleSelectionGetSelectedItemPropsOptions,
} from "downshift";
import type { LegacyRef } from "react";

/**
 * ------------------------------------------------------------------------------
 * Util function to get props for individual dropdown items
 * May conditionally call `getItemProps` or `getSelectedItemProps` depending on
 * whether `isMulti` is true, and whether the item is selected or not
 * ------------------------------------------------------------------------------
 */

interface GetDropdownItemPropsArgs
  extends Pick<
    DropdownMenuProps,
    "getItemProps" | "getSelectedItemProps" | "removeSelectedItem"
  > {
  item: DropdownItemShape;
  isItemSelected?: boolean;
  [key: string]: unknown;
}

export const getDropdownItemProps = ({
  isItemSelected,
  getItemProps,
  getSelectedItemProps,
  removeSelectedItem,
  item,
  isMulti,
  ...rest
}: GetDropdownItemPropsArgs) => {
  if (isMulti && getSelectedItemProps) {
    return isItemSelected
      ? {
          ...getItemProps({
            item,
            ...rest,
          }),
          ...getSelectedItemProps({
            selectedItem: item,
            onClick: (e) => {
              e.stopPropagation();
              if (removeSelectedItem) {
                removeSelectedItem(item);
              }
            },
          }),
        }
      : getItemProps({ item, ...rest });
  }

  return getItemProps({ item });
};

/**
 * ------------------------------------------------------------------------------
 * Renders a dropdown menu for use with `SelectSingle` or `SelectMultiple`
 * ------------------------------------------------------------------------------
 */

export interface DropdownMenuProps extends SprinklesArgs {
  /** Required props  */
  getIsItemSelected: (item: DropdownItemShape) => boolean;
  getItemProps: UseComboboxPropGetters<DropdownItemShape>["getItemProps"];
  getMenuProps: UseComboboxPropGetters<DropdownItemShape>["getMenuProps"];
  isOpen: boolean;
  items: Array<DropdownItemShape>;

  /** Optional props  */
  getSelectedItemProps?: (
    options: UseMultipleSelectionGetSelectedItemPropsOptions<DropdownItemShape>
  ) => { [key: string]: unknown };
  isMulti?: boolean;
  highlightedIndex?: number | undefined;
  removeSelectedItem?: UseMultipleSelectionActions<DropdownItemShape>["removeSelectedItem"];
  size?: SharedUiScale;
}

export const DropdownMenu = forwardRef(
  (
    {
      getIsItemSelected,
      getItemProps,
      getMenuProps,
      getSelectedItemProps,
      highlightedIndex,
      removeSelectedItem,
      isMulti,
      isOpen,
      items,
      size = "md",
      ...rest
    }: DropdownMenuProps,
    ref
  ) => {
    const dropdownWrapperStyles = clsx(styles.dropdownListWrapper, {
      [styles.dropdownWrapperClosed]: !isOpen,
    });

    const { atomProps, otherProps } = extractAtomsFromProps(rest, getSprinkles);

    /**
     * Note: `DropdownMenu` *must* not be in a conditional render, or
     * downshift's `getMenuProps` will be unable to apply a ref and throw an error
     */
    return (
      <Box
        className={dropdownWrapperStyles}
        {...atomProps}
        {...getMenuProps?.({
          ...otherProps,
          ref: ref as LegacyRef<HTMLElement>,
        })}
      >
        {Array.isArray(items) &&
          items.map((item, index) => {
            if (!item.label) {
              return null;
            }

            const isItemSelected = getIsItemSelected?.(item);

            return (
              <DropdownItem
                item={item}
                isMulti={isMulti}
                isHighlighted={highlightedIndex === index}
                key={item.label}
                isDropdownItemSelected={isItemSelected}
                size={size}
                {...getDropdownItemProps({
                  isItemSelected,
                  index,
                  getItemProps,
                  getSelectedItemProps,
                  item,
                  isMulti,
                  removeSelectedItem,
                })}
              />
            );
          })}
      </Box>
    );
  }
);
