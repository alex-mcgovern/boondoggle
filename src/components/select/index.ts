export { SelectSingle } from "./select_single.comp";
export type { SelectSingleProps } from "./select_single.comp";

export { SelectButton } from "./select_button.comp";
export type { SelectButtonProps } from "./select_button.comp";

export { SelectMulti } from "./select_multi.comp";
export type { SelectMultiProps } from "./select_multi.comp";

export type { DropdownItemShape } from "./select.types";
