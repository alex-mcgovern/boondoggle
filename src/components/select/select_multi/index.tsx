"use client";

import { autoUpdate, flip, offset, useFloating } from "@floating-ui/react";
import { faAngleDown } from "@fortawesome/pro-solid-svg-icons";
import clsx from "clsx";
import { useCombobox, useMultipleSelection } from "downshift";
import { forwardRef, useCallback, useState } from "react";

import { getOptionalLabelProps } from "../../../common-types";
import { useForwardRef } from "../../../hooks/use_forward_ref";
import { arrayHasLength } from "../../../lib/array_has_length";
import { Box } from "../../box";
import { Icon } from "../../icon";
import { Input } from "../../input";
import { SelectItemList } from "../SelectItemList";
import { filterSelectItems } from "../filterSelectItems";
import { flattenSelectItems } from "../flattenSelectItems";
import { getIsSelected } from "../getIsSelected";
import { getSlotRight } from "../lib/get_slot_right";
import { selectInputCursorStyles } from "../shared/select_input.styles.css";
import { selectMultiInputSelectedItemsStyle } from "./styles.css";

import type { UseComboboxStateChange } from "downshift";
import type { ForwardedRef } from "react";
import type {
	WithName,
	WithOptionalIsClearable,
	WithOptionalLabel,
	WithOptionalPlaceholder,
	WithSize,
	WithSlots,
	WithStateDisabled,
	WithStateInvalid,
} from "../../../common-types";
import { UtilCssArgs } from "../../../styles/utils/util_css";
import type { InputProps } from "../../input";
import type {
	FlatSelectItems,
	GroupedSelectItems,
	SelectItemShape,
} from "../types";

type GetPlaceholderArgs<
	TValue extends string = string,
	TItemData extends Record<string, unknown> = Record<string, unknown>,
> = {
	/**
	 * The placeholder text to display when no items are selected.
	 */
	placeholder: string | undefined;

	/**
	 * The selected items.
	 */
	selectedItems: Array<SelectItemShape<TValue, TItemData>> | undefined;

	/**
	 * A function that returns a string representation of the selected items.
	 */
	selectedItemsToString:
		| ((selectedItems: Array<SelectItemShape<TValue, TItemData>>) => string)
		| undefined;
};

/**
 * Returns the value to display in the input.
 */
function getPlaceholder<
	TValue extends string = string,
	TItemData extends Record<string, unknown> = Record<string, unknown>,
>({
	placeholder,
	selectedItems,
	selectedItemsToString,
}: GetPlaceholderArgs<TValue, TItemData>) {
	if (!arrayHasLength(selectedItems)) {
		return placeholder;
	}

	if (selectedItemsToString) {
		return selectedItemsToString(selectedItems);
	}

	return `${selectedItems.length} selected`;
}

export type SelectMultiProps<
	TValue extends string = string,
	TItemData extends Record<string, unknown> = Record<string, unknown>,
> = Partial<WithOptionalLabel> &
	WithStateInvalid &
	Omit<WithOptionalIsClearable, "readOnly"> &
	WithSlots &
	WithSize &
	WithStateDisabled &
	WithName &
	WithOptionalPlaceholder & {
		/**
		 * The bottom margin applied to the input element.
		 * @default "space_6" (1.5rem / 24px)
		 */
		marginBottom?: UtilCssArgs["marginBottom"];

		/**
		 * Item to be preselected when the component mounts.
		 */
		initialSelectedItems?: Array<SelectItemShape<TValue, TItemData>>;

		/**
		 * Props to customise the input element.
		 */
		inputProps?: Partial<InputProps>;

		/**
		 * Whether the Select should be filterable by typing.
		 */
		isFilterable?: boolean;

		/**
		 * Prop to toggle the open state of the dropdown.
		 */
		isOpen?: boolean;

		/**
		 * The items to render in the dropdown.
		 */
		items:
			| FlatSelectItems<TValue, TItemData>
			| GroupedSelectItems<TValue, TItemData>;

		/**
		 * Function called with the selected items when the selection changes.
		 */
		onChange?: (
			selection: Array<SelectItemShape<TValue, TItemData>> | undefined,
		) => void;

		/**
		 * Function called with the new open state when the dropdown is opened or closed.
		 */
		onIsOpenChange?: (
			changes: UseComboboxStateChange<SelectItemShape<TValue, TItemData>>,
		) => void;

		/**
		 * An array of selected items, used to control the component from outside.
		 */
		selectedItems?: Array<SelectItemShape<TValue, TItemData>>;

		/**
		 * A function that returns a string representation of the selected items.
		 */
		selectedItemsToString?: (
			selectedItems: Array<SelectItemShape<TValue, TItemData>>,
		) => string;
	};

function SelectMultiBase<
	TValue extends string = string,
	TItemData extends Record<string, unknown> = Record<string, unknown>,
>(
	{
		disabled,
		errorMessage,
		initialSelectedItems = [],
		invalid,
		isClearable,
		isFilterable,
		isLabelVisible,
		items: initialItems,
		label,
		labelTooltip,
		name,
		onChange,
		onIsOpenChange,
		placeholder,
		selectedItems: controlledSelectedItems,
		selectedItemsToString,
		size,
		slotLeft,
		slotRight = <Icon icon={faAngleDown} />,
		strClear,
		marginBottom = "space_6",
	}: SelectMultiProps<TValue, TItemData>,
	initialRef: ForwardedRef<HTMLInputElement>,
) {
	const ref = useForwardRef(initialRef);

	const [inputValue, setInputValue] = useState("");

	const items = isFilterable
		? filterSelectItems({
				inputValue,
				items: initialItems,
		  })
		: initialItems;

	const {
		addSelectedItem,
		getDropdownProps,
		getSelectedItemProps,
		removeSelectedItem,
		selectedItems,
	} = useMultipleSelection<SelectItemShape<TValue, TItemData>>({
		initialSelectedItems: controlledSelectedItems || [
			...initialSelectedItems,
			...flattenSelectItems(initialItems).filter((i) => {
				return i.isSelected;
			}),
		],
		onSelectedItemsChange: (c) => {
			return onChange?.(c.selectedItems);
		},
	});

	const getIsItemSelected = useCallback(
		(item: SelectItemShape<TValue, TItemData>) => {
			return getIsSelected({
				isMulti: true,
				item,
				selectedItems,
			});
		},
		[selectedItems],
	);

	const {
		getInputProps,
		getItemProps,
		getLabelProps,
		getMenuProps,
		highlightedIndex,
		isOpen,
		reset,
	} = useCombobox<SelectItemShape<TValue, TItemData>>({
		defaultHighlightedIndex: 0,
		isItemDisabled: (item) => {
			return item.disabled;
		},
		items: flattenSelectItems(items),
		onIsOpenChange, // after selection, highlight the first item.
		onStateChange({
			inputValue: newInputValue,
			selectedItem: newItem,
			type,
		}) {
			switch (type) {
				case useCombobox.stateChangeTypes.InputKeyDownEnter:
				case useCombobox.stateChangeTypes.ItemClick: {
					if (!newItem) {
						return;
					}

					newItem?.onClick?.();

					if (getIsItemSelected(newItem)) {
						removeSelectedItem(newItem);
					} else {
						addSelectedItem(newItem);
					}

					break;
				}

				case useCombobox.stateChangeTypes.InputChange:
					if (typeof newInputValue !== "undefined") {
						setInputValue(newInputValue);

						break;
					}

					break;

				default:
					break;
			}
		},
		selectedItem: null,
		stateReducer(_, actionAndChanges) {
			const { changes, type } = actionAndChanges;

			switch (type) {
				case useCombobox.stateChangeTypes.InputKeyDownEnter:
				case useCombobox.stateChangeTypes.ItemClick:
					return {
						...changes,
						highlightedIndex: 0, // with the first option highlighted.
						isOpen: true, // keep the menu open after selection.
					};

				default:
					return changes;
			}
		},
	});

	const { floatingStyles, refs } = useFloating({
		middleware: [
			offset(4),
			flip({
				crossAxis: true,
				fallbackAxisSideDirection: "start",
			}),
		],
		open: isOpen,
		placement: "bottom-start",
		whileElementsMounted: autoUpdate,
	});

	return (
		<Box position="relative">
			<Input
				errorMessage={errorMessage}
				invalid={invalid}
				outerRef={refs.setReference}
				size={size}
				slotLeft={slotLeft}
				marginBottom={marginBottom}
				slotRight={getSlotRight({
					isClearable:
						(!!isFilterable && !!inputValue) ||
						(!!isClearable && arrayHasLength(selectedItems)),
					reset,
					slotRight,
					strClear,
				})}
				{...getInputProps?.({
					className: clsx(selectInputCursorStyles, {
						[selectMultiInputSelectedItemsStyle]:
							arrayHasLength(selectedItems),
					}),
					disabled,
					id: name,
					placeholder: getPlaceholder({
						placeholder,
						selectedItems,
						selectedItemsToString,
					}),
					readOnly: !isFilterable,
					value: inputValue,
					...getDropdownProps({
						preventKeyAction: isOpen,
						ref,
					}),
					...getOptionalLabelProps({
						isLabelVisible,
						label,
						labelProps: getLabelProps({
							htmlFor: name,
						}),
						labelTooltip,
						name,
					}),
				})}
			/>
			<SelectItemList<TValue, TItemData>
				getIsItemSelected={getIsItemSelected}
				getItemProps={getItemProps}
				getMenuProps={getMenuProps}
				getSelectedItemProps={getSelectedItemProps}
				highlightedIndex={highlightedIndex}
				isMulti
				isOpen={isOpen}
				items={filterSelectItems({
					inputValue,
					items,
				})}
				ref={isOpen ? refs.setFloating : undefined}
				size={size}
				style={isOpen ? floatingStyles : {}}
			/>
		</Box>
	);
}

export const SelectMulti = forwardRef(SelectMultiBase);
