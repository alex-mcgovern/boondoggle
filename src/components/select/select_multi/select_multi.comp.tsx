import { useCombobox, useMultipleSelection } from "downshift";
import { forwardRef, useCallback, useMemo, useState } from "react";

import { Box } from "../../box";
import { Input } from "../../input";
import {
  downshiftStateReducer,
  getDefaultHighlightedIndex,
  getDisplayValue,
  getFilteredDropdownItems,
  getIsSelected,
} from "../select_utils";
import { DEFAULT_SLOT_RIGHT } from "../shared/DEFAULT_SLOT_RIGHT";
import { DropdownMenu } from "../shared/dropdown_menu/dropdown_menu.comp";
import { selectInputCursorStyles } from "../shared/select_input.styles.css";

import type { LabelledElementCustomisation } from "../../../types";
import type { DropdownItemShape, SelectCommonProps } from "../select.types";
import type { UseMultipleSelectionStateChange } from "downshift";
import type { Ref } from "react";

export type SelectMultiProps = SelectCommonProps &
  LabelledElementCustomisation & {
    initialSelectedItems?: Array<DropdownItemShape>;
    itemToString?: (item: DropdownItemShape | null) => string;
    onChange?: (
      changes: UseMultipleSelectionStateChange<DropdownItemShape>
    ) => void;

    placeholder: string;
  };

export const SelectMulti = forwardRef(
  (
    {
      disabled,
      errorMessage,
      id,
      initialHighlightedItem,
      initialSelectedItems = [],
      invalid,
      isFilterable,
      items,
      label,
      name,
      onChange,
      placeholder,
      size,
      slotLeft,
      wrapperProps,
      slotRight = DEFAULT_SLOT_RIGHT,
      ...rest
    }: SelectMultiProps,
    ref: Ref<HTMLInputElement>
  ) => {
    /** Externally controlled state for downshift and {@link Input} component */
    const [inputValue, setInputValue] = useState("");
    const [inputPlaceholder, setInputPlaceholder] = useState(
      getDisplayValue({
        length: initialSelectedItems?.length,
        originalValue: placeholder,
      })
    );

    /**
     * Callback to update input value and call `onChange` (if provided) when selected items change.
     */
    const onSelectedItemsChange = useCallback(
      (changes: UseMultipleSelectionStateChange<DropdownItemShape>) => {
        if (onChange) {
          onChange(changes);
        }
        setInputPlaceholder(
          getDisplayValue({
            length: changes?.selectedItems?.length,
            originalValue: placeholder,
          })
        );
      },
      [onChange, placeholder]
    );

    /**
     * Downshift `useMultipleSelection` hook
     * @see https://www.downshift-js.com/use-multiple-selection
     */
    const {
      getSelectedItemProps,
      getDropdownProps,
      removeSelectedItem,
      selectedItems,
      setSelectedItems,
    } = useMultipleSelection<DropdownItemShape>({
      initialSelectedItems,
      onSelectedItemsChange,
    });

    /**
     * When `isFilterable` is true, we need to filter the items based on the input value.
     */
    const filteredItems = useMemo(() => {
      if (!items || !isFilterable) {
        return items;
      }

      return getFilteredDropdownItems({ inputValue, items });
    }, [items, isFilterable, inputValue]);

    /** -----------------------------------------------------------------------------
     * Util function for checking if item is selected, passed down to DropdownItemShape via DropdownMenu
     * ------------------------------------------------------------------------------- */

    const getIsItemSelected = useCallback(
      (item: DropdownItemShape) => {
        return getIsSelected({
          isMulti: true,
          item,
          selectedItems,
        });
      },
      [selectedItems]
    );

    /**
     * Downshift `useCombobox` hook
     * @see https://www.downshift-js.com/use-combobox
     */
    const {
      isOpen,
      toggleMenu,
      getMenuProps,
      getInputProps,
      highlightedIndex,
      getItemProps,
      getLabelProps,
    } = useCombobox<DropdownItemShape>({
      defaultHighlightedIndex: getDefaultHighlightedIndex({
        initialHighlightedItem,
        items,
      }),
      items: filteredItems,
      onStateChange({
        inputValue: newInputValue,
        type,
        selectedItem: newSelectedItem,
      }) {
        switch (type) {
          case useCombobox.stateChangeTypes.InputKeyDownEnter:
          case useCombobox.stateChangeTypes.ItemClick:
          case useCombobox.stateChangeTypes.InputBlur:
            if (!newSelectedItem) {
              break;
            }

            if (getIsItemSelected(newSelectedItem)) {
              removeSelectedItem(newSelectedItem);
              break;
            }

            if (newSelectedItem) {
              setSelectedItems([...selectedItems, newSelectedItem]);
              break;
            }

            break;

          case useCombobox.stateChangeTypes.InputChange:
            if (typeof newInputValue !== "undefined") {
              setInputValue(newInputValue);
              break;
            }

            break;
          default:
            break;
        }
      },
      selectedItem: null,
      stateReducer: (state, actionAndChanges) => {
        return downshiftStateReducer(state, actionAndChanges, {
          isMulti: true,
        });
      },
    });

    return (
      <Box {...wrapperProps}>
        <DropdownMenu
          errorMessage={errorMessage}
          getIsItemSelected={getIsItemSelected}
          getItemProps={getItemProps}
          getMenuProps={getMenuProps}
          getSelectedItemProps={getSelectedItemProps}
          highlightedIndex={highlightedIndex}
          isMulti
          isOpen={isOpen}
          items={filteredItems}
          removeSelectedItem={removeSelectedItem}
          size={size}
          width="100%"
          triggerNode={
            // eslint-disable-next-line react-perf/jsx-no-jsx-as-prop
            <Input
              className={selectInputCursorStyles}
              invalid={invalid}
              readOnly={!isFilterable}
              size={size}
              slotLeft={slotLeft}
              slotRight={slotRight}
              labelProps={getLabelProps({
                htmlFor: id,
                label,
              })}
              {...getInputProps?.({
                ...getDropdownProps({
                  preventKeyAction: isOpen,
                  ref,
                }),
                ...rest,
                disabled,
                id,
                name,
                onClick: toggleMenu,
                placeholder: inputPlaceholder,
                value: inputValue,
              })}
            />
          }
        />
      </Box>
    );
  }
);
