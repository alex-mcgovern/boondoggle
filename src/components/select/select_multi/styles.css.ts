import { globalStyle, style } from "@vanilla-extract/css";

import { vars } from "../../../styles/theme.css";

export const selectMultiInputSelectedItemsStyle = style({});

globalStyle(`${selectMultiInputSelectedItemsStyle}::placeholder`, {
    color: vars.color.text_high_contrast,
});
