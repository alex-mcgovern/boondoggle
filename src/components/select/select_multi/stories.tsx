import { useState } from "react";

import { SelectMulti as StoryComp } from ".";
import { LOREM } from "../../../../mocks/LOREM.mock";
import { Box } from "../../box/_components/Box";
import { mockSelectItems } from "../__mocks__/select.mock";

import type { Meta, StoryObj } from "@storybook/react";
import type { SelectItemShape } from "../types";

const ITEMS = mockSelectItems({});

const meta = {
	args: {
		errorMessage: LOREM.errorMessage(),
		items: ITEMS,
		label: LOREM.label(),
		name: LOREM.name(),
		// onChange: (selection) => {
		//     console.log(`onChange\n\n${JSON.stringify(selection, null, 2)}`);
		// },
		placeholder: LOREM.placeholder(),
	},
	component: StoryComp,
	title: "Components/SelectMulti",
} satisfies Meta<typeof StoryComp>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {};

export const Filterable: Story = {
	args: {
		isFilterable: true,
	},
};

export const InitialSelectedItems: Story = {
	args: {
		initialSelectedItems: [ITEMS[0]],
	},
};

export const Invalid: Story = {
	args: {
		invalid: true,
	},
};

export const ItemsWithIsSelectedTrue: Story = {
	args: {
		items: [
			{
				isSelected: true,
				label: "United Kingdom",
				value: "uk",
			},
			{
				isSelected: true,
				label: "France",
				value: "fr",
			},
			{
				isSelected: true,
				label: "Germany",
				value: "de",
			},
		],
	},
};

const ControlledTemplate = () => {
	const [selectedItems, setSelectedItems] = useState<
		Array<SelectItemShape<string>>
	>([]);

	return (
		<Box>
			<Box
				alignItems="center"
				display="flex"
				gap="space_2"
				marginBottom="space_4"
			>
				<button
					onClick={() => {
						return setSelectedItems([]);
					}}
					type="button"
				>
					Clear selected Items
				</button>
				<button
					onClick={() => {
						return setSelectedItems([ITEMS[0]]);
					}}
					type="button"
				>
					Set selected Items
				</button>
			</Box>
			<StoryComp
				items={ITEMS}
				name={LOREM.name()}
				placeholder={LOREM.placeholder()}
				selectedItems={selectedItems}
			/>
			{JSON.stringify(selectedItems)}
		</Box>
	);
};

export const Controlled: Story = {
	render: ControlledTemplate,
};

export const Customisation: Story = {
	args: {
		marginBottom: "none",
	},
};

export const Disabled: Story = {
	args: {
		disabled: true,
	},
};

export const OnChange: Story = {
	args: {
		onChange: (selectedItems) => {
			alert(JSON.stringify({ selectedItems }));
		},
	},
};

export const SelectedItemsToString: Story = {
	args: {
		selectedItemsToString: (
			selectedItems: Array<SelectItemShape<string>>,
		) => {
			if (selectedItems.length === 1) {
				return `${selectedItems.length} country selected`;
			}

			return `${selectedItems.length} countries selected`;
		},
	},
};

export const SlotLeft: Story = {
	args: {
		items: mockSelectItems({ withIcon: true }),
	},
};
