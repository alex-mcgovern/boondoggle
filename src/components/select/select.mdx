import { ArgTypes, Canvas, Meta, Story } from "@storybook/addon-docs";

import * as stories from "./select.stories.tsx";

<Meta of={stories} />

# Select

- Wrapper around Downshift to provide a select component with a few options for
  styling and behavior

## Select Single

<Canvas name="default" of={stories.SelectSingleDefault} />

## Select multiple

<Canvas name="default" of={stories.SelectMultipleDefault} />

## Filterable

[ToDo: Add filterable example]
{/* <Canvas name="filterable" of={stories.Filterable} /> */}
