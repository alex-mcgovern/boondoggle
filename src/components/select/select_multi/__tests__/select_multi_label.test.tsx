/** @jest-environment jsdom */
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { LOREM } from "../../../../../mocks/LOREM.mock";
import "../../../../../test/dialog.mock";
import { mockSelectItems } from "../../__mocks__/select.mock";
import { SelectMulti } from "../select_multi.comp";

import type { SelectMultiProps } from "../select_multi.comp";

/** ----------------------------------------------------------------------------- */

const PROPS: SelectMultiProps = {
  id: LOREM.id(),
  items: mockSelectItems({}),
  label: LOREM.label(),
  name: LOREM.text_xxs,
  placeholder: LOREM.select,
};

const renderComponent = (props: SelectMultiProps) => {
  return {
    user: userEvent.setup(),
    ...render(<SelectMulti {...props} />),
  };
};

/** ----------------------------------------------------------------------------- */

describe("<SelectMulti />", () => {
  it("should be labelled", async () => {
    const { container, getByRole } = await renderComponent(PROPS);

    const combobox = getByRole("combobox");
    const label = container.querySelector("label");
    const labelId = label?.getAttribute("id");

    expect(combobox.getAttribute("aria-labelledby")).toBe(labelId);
    expect(label).not.toBeNull();
    expect(label?.getAttribute("for")).toBe(PROPS.id);
    expect(label?.textContent).toBe(PROPS.label);
  });

  it("should match snapshot", async () => {
    const { getByRole } = await renderComponent(PROPS);

    expect(getByRole("combobox")).toMatchSnapshot();
  });
});
