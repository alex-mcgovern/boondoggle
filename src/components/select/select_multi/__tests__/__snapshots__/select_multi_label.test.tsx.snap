// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectMulti /> should match snapshot 1`] = `
<input
  aria-activedescendant=""
  aria-autocomplete="list"
  aria-controls="downshift-2-menu"
  aria-expanded="false"
  aria-labelledby="downshift-2-label"
  autocomplete="off"
  class="slot_wrapper_inset.styles_offsetSlotRightStyle__1r5w8ord input.styles_getInputStyles__3x32wp9 input.styles__3x32wp8 get_sprinkles_width_100%_mobile__eiij7bgk a11y_a11yDisabled__r8he5t1 input.styles_getInputStyles_hasBorder_true__3x32wpb input.styles_hasBorder_true__3x32wp7 input.styles_getInputStyles_size_md__3x32wpd input.styles_size_md__3x32wp4 element_size_elementFontSize_md__1f6zrv4 element_size__1f6zrv1 get_sprinkles_fontSize_body_md__eiij7b20 get_sprinkles_lineHeight_body_md__eiij7b3e element_size_elementPadding_md__1f6zrva element_size__1f6zrv7 get_sprinkles_paddingLeft_spacing_1.5_mobile__eiij7bdw get_sprinkles_paddingRight_spacing_1.5_mobile__eiij7bet get_sprinkles_paddingTop_spacing_1_mobile__eiij7bfn get_sprinkles_paddingBottom_spacing_1_mobile__eiij7bcw input.styles__3x32wp1 get_sprinkles_height_spacing_5__eiij7b32 select_input.styles_selectInputCursorStyles__1bkra5a0"
  id="lorem-ipsum"
  name="Lorem ipsum"
  placeholder="Select a thing"
  readonly=""
  role="combobox"
  value=""
/>
`;
